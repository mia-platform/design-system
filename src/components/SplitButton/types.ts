import { DropdownItem } from '../Dropdown/props'

export type SplitButtonItem = Pick<DropdownItem, 'icon'|'id'|'label'>
