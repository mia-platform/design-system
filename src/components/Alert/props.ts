/**
 * Copyright 2024 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { ReactNode } from 'react'

import { IconComponent } from '../Icon/Icon.props.ts'
import { Type } from './types.ts'

export type AlertProps = {

  /**
   * The title displayed at the top of the alert.
   */
  title?: ReactNode

  /**
   * The description providing additional details about the alert.
   */
  description?: ReactNode

  /**
   * Custom children content to be rendered inside the alert.
   */
  children?: ReactNode

  /**
   * Indicates if the alert is closable.
   */
  isClosable?: boolean

  /**
   * The type of alert.
   */
  type?: Type

  /**
   * If true, renders a compact version of the alert with reduced padding.
   */
  isCompressed?: boolean

  /**
   * A custom icon to display in the alert. Replaces the default icon based on the `type`.
   */
  icon?: IconComponent

  /**
   * A custom action element (e.g., a button) to display inside the alert.
   */
  action?: ReactNode

  /**
   * A callback function triggered when the alert is closed.
   */
  onClose?: () => void
}
