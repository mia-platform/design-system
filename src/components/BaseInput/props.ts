import { InputRef } from 'antd'
import { Ref } from 'react'

import { Appearance } from './types.ts'

export type BaseInputProps = {

  /**
   * The placeholder for the input.
   */
  placeholder?: string

  /**
   * The type of the Input.
   */
  appearance?: Appearance

  /**
   * Whether the input is disabled.
   */
  isDisabled?: boolean,

  /**
   * Whether the input has error.
   */
  isError?: boolean

  /**
   * Whether the input is read-only.
   */
  isReadOnly?: boolean

  /**
   * Whether the input is full width. Defaults to true.
   */
  isFullWidth?: boolean

  /**
   * Ref to the antd input.
   */
  inputRef?: Ref<InputRef>

  /**
   * Classes for the input.
   */
  className?: string

}
