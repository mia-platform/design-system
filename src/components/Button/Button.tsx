/**
 * Copyright 2023 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { ReactElement, ReactNode, useMemo } from 'react'
import { Button as AntButton } from 'antd'
import classnames from 'classnames'

import { ButtonHierarchies, ButtonIconPositions, ButtonShapes, ButtonSizes, ButtonTypes } from './Button.types'
import styles from './Button.module.css'

const { button, buttonSm, buttonSmIconOnly, buttonMd, buttonMdIconOnly, buttonLg, buttonGhost, buttonText } = styles

const { Primary, Neutral, Danger } = ButtonHierarchies
const { Left, Right } = ButtonIconPositions
const { Square } = ButtonShapes
const { Small, Middle, Large } = ButtonSizes
const { Filled, Ghost } = ButtonTypes

export type ButtonProps = {

  /**
   * The children nodes to be rendered within the button context.
   */
  children?: ReactNode,

  /**
   * Defines the button hierarchy. Either:
   *
   * - primary: button associated with the most significant (and therefore primary) action on the page;
   * - neutral: button associated with a general purpose action;
   * - danger: button associated with a potentially dangerous action.
   */
  hierarchy?: ButtonHierarchies,

  /**
   * Redirect url of a link button.
   * For security reasons, the attribute "rel: 'noopener noreferrer'" is always specified.
   */
  href?: string,

  /**
   * Sets an icon for the button component.
   */
  icon?: ReactNode,

  /**
   * Defines a position for the button icon (if any). Either:
   *
   * - left: the icon is placed to the left of the button text;
   * - right: the icon is placed to the right of the button text.
   */
  iconPosition?: ButtonIconPositions,

  /**
   * Identifies whether the button is disabled or not.
   */
  isDisabled?: boolean,

  /**
   * Sets the loading status of the button.
   */
  isLoading?: boolean,

  /**
   * Sets the handler to handle a click event.
   */
  onClick?: () => void,

  /**
   * Defines the button shape. Either:
   *
   * - square: square-shaped button with rounded corners;
   * - circle: fully rounded button, typically used for buttons containing only icon.
   */
  shape?: ButtonShapes,

  /**
   * Defines the button size. Either:
   *
   * - small: small button - 24px in height by default;
   * - middle: middle button - 32px in height by default;
   * - large: large button - 48px in height by default.
   */
  size?: ButtonSizes,

  /**
   * Specifies where the linked document will open when the link is clicked.
   * Only usable for link buttons together with the href property.
   *
   * @link https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a
   */
  target?: string,

  /**
   * Defines the button type. Either:
   *
   * - filled: button with a solid fill, typically used for primary actions;
   * - outlined: medium-emphasis button with a stroke around the button container and no fill,
   *   typically used for secondary actions;
   * - ghost: transparent button with no stroke, typically used for links.
   */
  type?: ButtonTypes,
}

/**
 * UI component for performing actions on the page interacting through clicks
 *
 * @link https://ant.design/components/button
 * @returns {Button} Button component
 */
export const Button = ({
  children,
  hierarchy,
  href,
  icon,
  iconPosition,
  isDisabled,
  isLoading,
  onClick,
  shape,
  size,
  target,
  type,
}: ButtonProps): ReactElement => {
  const buttonClassNames = useMemo(() => classnames(
    [
      button,
      children && size === Small && buttonSm,
      !children && size === Small && buttonSmIconOnly,
      children && size === Middle && buttonMd,
      !children && size === Middle && buttonMdIconOnly,
      size === Large && buttonLg,
      type === Ghost && buttonGhost,
    ]
  ), [children, size, type])

  return (
    <AntButton
      className={buttonClassNames}
      danger={hierarchy === Danger}
      disabled={isDisabled}
      ghost={type !== Filled && hierarchy !== Neutral}
      loading={isLoading}
      shape={shape === Square ? 'default' : 'circle'}
      size={size}
      type={hierarchy === Neutral ? 'default' : 'primary'}
      onClick={onClick}
      {...href && { href, rel: 'noopener noreferrer', target }}
    >
      {iconPosition === Left && icon}
      {children && <div className={buttonText}>{children}</div>}
      {iconPosition === Right && icon}
    </AntButton>
  )
}

Button.defaultProps = {
  hierarchy: Primary,
  iconPosition: Left,
  isDisabled: false,
  isLoading: false,
  shape: Square,
  size: Middle,
  type: Filled,
}
