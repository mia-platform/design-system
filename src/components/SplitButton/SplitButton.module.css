/**
 * Copyright 2023 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

.splitButton {
  display: flex;
}

.mainActionButton {
  border-top-right-radius: 0px;
  border-bottom-right-radius: 0px;

  &:hover {
    z-index: 999;
  }

  &.outlined {
    /*
     * Transition does not affect the border in the middle of the two buttons
     * because its visibility is controlled by the z-index attribute.
     * This means that only three out of four borders would be properly animated.
     * Since the issue is only visible in outlined buttons we keep other variants animated.
     */
    transition: none
  }
}

.dropdownActionButton {
  margin-left: 1px;
  border-top-left-radius: 0px;
  border-bottom-left-radius: 0px;

  &.outlined, &:disabled {
    margin-left: -1px;
    /*
     * Transition is removed for coherence with the main button.
     */
    transition: none
  }

  &.pressed {
    background-color: var(--palette-action-primary-selected, #004aaa);
  }
}
