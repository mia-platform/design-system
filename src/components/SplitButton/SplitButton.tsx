/**
 * Copyright 2023 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { ReactElement, useCallback, useState } from 'react'
import { PiCaretDown } from 'react-icons/pi'
import classNames from 'classnames'

import { Button } from '../Button/Button'
import { Dropdown } from '../Dropdown/Dropdown'
import { Icon } from '../Icon'
import { SplitButtonProps } from './props'
import styles from './SplitButton.module.css'

const {
  dropdownActionButton,
  mainActionButton,
  outlined,
  pressed,
  splitButton,
} = styles

const dropdownIcon = <Icon aria-label="Open dropdown" component={PiCaretDown} size={16} />

export const defaults = {
  hierarchy: Button.Hierarchy.Primary,
}

export const SplitButton = ({
  autoFocus,
  children,
  hierarchy = defaults.hierarchy,
  href,
  htmlType,
  isDisabled,
  isLoading,
  items,
  itemLayout,
  onClick,
  onItemClick,
  target,
  title,
  type,
}: SplitButtonProps): ReactElement => {
  const [opened, setOpened] = useState(false)
  const onOpenChange = useCallback(() => setOpened((prev) => !prev), [])

  return (
    <div className={splitButton}>
      <Button
        autoFocus={autoFocus}
        className={classNames(
          mainActionButton,
          type === Button.Type.Outlined || hierarchy === Button.Hierarchy.Neutral ? outlined : undefined
        )}
        hierarchy={hierarchy}
        htmlType={htmlType}
        isDisabled={isDisabled}
        isLoading={isLoading}
        title={title}
        type={type}
        onClick={onClick}
        {...href && { href, rel: 'noopener noreferrer', target }}
      >
        {children}
      </Button>
      <Dropdown
        itemLayout={itemLayout}
        items={items}
        placement={Dropdown.Placement.BottomRight}
        onClick={onItemClick}
        onOpenChange={onOpenChange}
      >
        <Button
          className={classNames(
            dropdownActionButton,
            type === Button.Type.Outlined || hierarchy === Button.Hierarchy.Neutral ? outlined : undefined,
            opened ? pressed : undefined
          )}
          hierarchy={hierarchy}
          icon={dropdownIcon}
          isDisabled={isDisabled}
          type={type}
        />
      </Dropdown>
    </div>
  )
}
