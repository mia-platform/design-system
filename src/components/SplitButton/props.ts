/**
 * Copyright 2024 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { MouseEvent, ReactNode } from 'react'

import { DropdownClickEvent, ItemLayout } from '../Dropdown/props'
import { HTMLType, Hierarchy, Type } from '../Button/Button.types'
import { SplitButtonItem } from './types'

export type SplitButtonProps = {

  /**
   * Use this to focus the dropdown element.
   */
  autoFocus?: boolean

  /**
   * The children nodes to be rendered within the button context.
   */
  children?: ReactNode,

  /**
   * Defines the button hierarchy. Either:
   *
   * - "primary": button associated with the most significant (and therefore primary) action on the page;
   * - "neutral": button associated with a general purpose action;
   * - "danger": button associated with a potentially dangerous action.
   */
  hierarchy?: Hierarchy,

  /**
   * Redirect url of a link button.
   * For security reasons, the attribute "rel: 'noopener noreferrer'" is always specified.
   */
  href?: string,

  /**
   * The underlying html button type. Either:
   *
   * - "button"
   * - "reset"
   * - "submit"
   */
  htmlType?: HTMLType

  /**
   * Identifies whether the button is disabled or not.
   */
  isDisabled?: boolean,

  /**
   * Sets the loading status of the button.
   */
  isLoading?: boolean,

  /**
   * Allows to control the Dropdown label layout (accepts: horizontal, vertical).
   */
  itemLayout?: ItemLayout,

  /**
   * List of items to be rendered within the Dropdown.
   *
   * Note: This component currently supports items with vertical layout
   */
  items: SplitButtonItem[]

  /**
   * callback used to notify a click on the main button.
   */
  onClick: (event: MouseEvent) => void,

  /**
   * callback used to notify a click on a dropdown item, see
   * Dropdown#onClick specification for further details.
   */
  onItemClick: (event: DropdownClickEvent) => void

  /**
   * Specifies where the linked document will open when the link is clicked.
   * Only usable for link buttons together with the href property.
   *
   * @link https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a
   */
  target?: string,

  type?: Type,

  /**
   * HTML `title` global attribute.
   */
  title?: string
}
