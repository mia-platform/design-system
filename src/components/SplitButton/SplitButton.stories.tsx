/**
 * Copyright 2024 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import type { Meta, StoryObj } from '@storybook/react'
import { PiAcorn, PiWarning } from 'react-icons/pi'
import { action } from '@storybook/addon-actions'

import { Button } from '../Button'
import { Icon } from '../Icon'
import { SplitButton } from './SplitButton'
import { SplitButtonProps } from './props'
import { Tag } from '../Tag'
import { Dropdown } from '../Dropdown'

const defaults: Partial<SplitButtonProps> = {
  children: 'click me',
  items: [{
    icon: <Icon component={PiAcorn} size={16} />,
    id: 'id1',
    label: 'Action 1',
  }, {
    id: 'id2',
    label: 'Action 2',
  }],
  onClick: action('on click'),
  onItemClick: action('on item click'),
}

const meta = {
  component: SplitButton,
  args: defaults,
  argTypes: {},
  render: (_, { args }) => (
    <div
      style={{
        display: 'flex',
        justifyContent: 'center',
        alignItems: 'center',
      }}
    >
      <SplitButton {...args} />
    </div>
  ),
} satisfies Meta<typeof SplitButton>

type Story = StoryObj<typeof meta>

export default meta

export const BasicExample: Story = {}

export const PrimaryFilled: Story = {
  args: {
    hierarchy: Button.Hierarchy.Primary,
  },
}
export const PrimaryOutlined: Story = {
  args: {
    hierarchy: Button.Hierarchy.Primary,
    type: Button.Type.Outlined,
  },
}
export const Neutral: Story = {
  args: {
    hierarchy: Button.Hierarchy.Neutral,
  },
}
export const Danger: Story = {
  args: {
    hierarchy: Button.Hierarchy.Danger,
  },
}

export const Disabled: Story = {
  args: {
    isDisabled: true,
  },
}

export const Loading: Story = {
  args: {
    isLoading: true,
  },
}

export const WithComplexMenu: Story = {
  args: {
    items: [{
      icon: <Icon component={PiAcorn} size={16} />,
      id: 'id1',
      label: 'value 1',
    }, {
      id: 'id2',
      label: 'value 2',
      secondaryLabel: 'Some additional info 2',
    }, {
      icon: <Icon component={PiWarning} size={16} />,
      id: 'id3',
      label: 'I am danger!',
      danger: true,
      secondaryLabel: 'Some additional info 3',
    }, {
      id: 'id4-with-nested-items',
      label: 'Many Items here',
      secondaryLabel: 'just hover me',
      children: [{
        id: 'id4-c1',
        label: 'value 4.1',
      }, {
        id: 'id4-c2',
        label: 'value 4.2',
        children: [{
          id: 'id4-c2-c1',
          label: 'value 4.2-1',
        }],
      }],
    }, {
      id: 'id5-with-tag',
      label: 'with tag',
      secondaryLabel: 'Some additional info 2',
      tag: <Tag>{'Tag'}</Tag>,
    }],
  },
}

export const WithComplexMenuAndVerticalLayout: Story = {
  args: {
    itemLayout: Dropdown.ItemLayout.Vertical,
    items: [{
      icon: <Icon component={PiAcorn} size={16} />,
      id: 'id1',
      label: 'value 1',
    }, {
      id: 'id2',
      label: 'value 2',
      secondaryLabel: 'Some additional info 2',
    }, {
      icon: <Icon component={PiWarning} size={16} />,
      id: 'id3',
      label: 'I am danger!',
      danger: true,
      secondaryLabel: 'Some additional info 3',
    }, {
      id: 'id4-with-nested-items',
      label: 'Many Items here',
      secondaryLabel: 'just hover me',
      children: [{
        id: 'id4-c1',
        label: 'value 4.1',
      }, {
        id: 'id4-c2',
        label: 'value 4.2',
        children: [{
          id: 'id4-c2-c1',
          label: 'value 4.2-1',
        }],
      }],
    }, {
      id: 'id5-with-tag',
      label: 'with tag',
      secondaryLabel: 'Some additional info 2',
      tag: <Tag>{'Tag'}</Tag>,
    }],
  },
}
