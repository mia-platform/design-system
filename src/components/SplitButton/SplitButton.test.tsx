/**
 * Copyright 2024 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { RenderResult, render as renderComponent, screen, userEvent } from '../../test-utils'
import { Button } from '../Button'
import { SplitButton } from './SplitButton'
import { SplitButtonItem } from './types'
import { SplitButtonProps } from './props'

const items: SplitButtonItem[] = [
  { id: '1', label: 'Label 1' },
  { id: '2', label: 'Label 2' },
]
const defaultProps: SplitButtonProps = {
  items,
  children: <span>{'test-trigger-button'}</span>,
  onClick: jest.fn(),
  onItemClick: jest.fn(),
}

describe('SplitButton Component', () => {
  beforeEach(jest.clearAllMocks)

  describe('ui variants', () => {
    it('renders primary filled', () => {
      const { baseElement } = render({ props: { ...defaultProps, hierarchy: Button.Hierarchy.Primary } })
      expect(baseElement).toMatchSnapshot()
    })

    it('renders primary outlined', () => {
      const { baseElement } = render({ props: {
        ...defaultProps,
        hierarchy: Button.Hierarchy.Primary,
        type: Button.Type.Outlined,
      } })
      expect(baseElement).toMatchSnapshot()
    })

    it('renders neutral (outlined)', () => {
      const { baseElement } = render({ props: {
        ...defaultProps,
        hierarchy: Button.Hierarchy.Neutral,
      } })
      expect(baseElement).toMatchSnapshot()
    })

    it('renders danger (filled)', () => {
      const { baseElement } = render({ props: {
        ...defaultProps,
        hierarchy: Button.Hierarchy.Danger,
      } })
      expect(baseElement).toMatchSnapshot()
    })

    it('renders danger (outlined)', () => {
      const { baseElement } = render({ props: {
        ...defaultProps,
        hierarchy: Button.Hierarchy.Danger,
        type: Button.Type.Outlined,
      } })
      expect(baseElement).toMatchSnapshot()
    })
  })

  describe('href button', () => {
    it('correctly sets rel attributes', () => {
      render({ props: { ...defaultProps, href: 'some-url.com' } })

      const link = screen.getByRole('link')
      expect(link).toBeInTheDocument()
      expect(link.getAttribute('rel')).toEqual('noopener noreferrer')
    })
  })

  describe('main button', () => {
    it('renders child node', () => {
      render()
      expect(screen.getByText('test-trigger-button')).toBeInTheDocument()
    })

    it('invokes onClick', async() => {
      const onClick = jest.fn()
      render({ props: { ...defaultProps, onClick } })
      const mainButton = screen.getByRole('button', { name: 'test-trigger-button' })
      expect(mainButton).toBeInTheDocument()

      await userEvent.click(mainButton)
      expect(onClick).toHaveBeenCalledTimes(1)
    })
  })

  describe('dropdown button', () => {
    it('renders secondary button', () => {
      render()
      expect(screen.getAllByRole('button')).toHaveLength(2)
      expect(screen.getByRole('button', { name: /open dropdown/i })).toBeInTheDocument()
    })

    it('opens dropdown menu on secondary button click', async() => {
      const onClick = jest.fn()
      const onItemClick = jest.fn()
      render({ props: { ...defaultProps, onClick, onItemClick } })

      const dropdownButton = screen.getByRole('button', { name: /open dropdown/i })
      expect(dropdownButton).toBeInTheDocument()

      await userEvent.click(dropdownButton)
      const firstMenuItem = await screen.findByRole('menuitem', { name: 'Label 1' })

      expect(screen.getAllByRole('menuitem')).toHaveLength(2)

      await userEvent.click(firstMenuItem)

      expect(onClick).not.toHaveBeenCalled()
      expect(onItemClick).toHaveBeenCalled()
      expect(onItemClick).toHaveBeenCalledTimes(1)
      expect(onItemClick).toHaveBeenCalledWith({
        id: '1',
        item: { id: '1', label: 'Label 1' },
        selectedPath: ['1'],
        domEvent: expect.anything(),
      })
    })

    it('calls onOpenChange when dropdown is triggered', async() => {
      const onOpenChange = jest.fn()
      render({ props: { ...defaultProps, onOpenChange } })

      const dropdownButton = screen.getByRole('button', { name: /open dropdown/i })
      expect(dropdownButton).toBeInTheDocument()

      await userEvent.click(dropdownButton)
      const firstMenuItem = await screen.findByRole('menuitem', { name: 'Label 1' })

      expect(screen.getAllByRole('menuitem')).toHaveLength(2)

      await userEvent.click(firstMenuItem)

      expect(onOpenChange).toHaveBeenCalledTimes(2)
      expect(onOpenChange).toHaveBeenNthCalledWith(1, true, { source: 'trigger' })
      expect(onOpenChange).toHaveBeenNthCalledWith(2, false, { source: 'menu' })
    })
  })
})

function render(
  { props }: { props: SplitButtonProps } = { props: defaultProps }
): RenderResult {
  return renderComponent(<SplitButton {...props} />)
}
