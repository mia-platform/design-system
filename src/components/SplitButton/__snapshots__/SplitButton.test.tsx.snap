// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SplitButton Component ui variants renders danger (filled) 1`] = `
<body>
  <div>
    <div
      class="splitButton"
    >
      <button
        class="mia-platform-btn mia-platform-btn-primary mia-platform-btn-dangerous mia-platform-btn-color-dangerous mia-platform-btn-variant-solid button buttonMd mainActionButton"
        type="button"
      >
        <div
          class="buttonText"
        >
          <span>
            test-trigger-button
          </span>
        </div>
      </button>
      <span
        class="mia-platform-dropdown-trigger dropdown-some-random-uuid"
      >
        <button
          class="mia-platform-btn mia-platform-btn-primary mia-platform-btn-dangerous mia-platform-btn-color-dangerous mia-platform-btn-variant-solid mia-platform-btn-icon-only button buttonMdIconOnly dropdownActionButton"
          type="button"
        >
          <span
            class="mia-platform-btn-icon"
          >
            <svg
              aria-label="Open dropdown"
              color="currentColor"
              fill="currentColor"
              height="16"
              role="img"
              stroke="currentColor"
              stroke-width="0"
              style="color: currentColor;"
              viewBox="0 0 256 256"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M213.66,101.66l-80,80a8,8,0,0,1-11.32,0l-80-80A8,8,0,0,1,53.66,90.34L128,164.69l74.34-74.35a8,8,0,0,1,11.32,11.32Z"
              />
            </svg>
          </span>
        </button>
      </span>
    </div>
  </div>
</body>
`;

exports[`SplitButton Component ui variants renders danger (outlined) 1`] = `
<body>
  <div>
    <div
      class="splitButton"
    >
      <button
        class="mia-platform-btn mia-platform-btn-primary mia-platform-btn-dangerous mia-platform-btn-color-dangerous mia-platform-btn-variant-solid mia-platform-btn-background-ghost button buttonMd mainActionButton outlined"
        type="button"
      >
        <div
          class="buttonText"
        >
          <span>
            test-trigger-button
          </span>
        </div>
      </button>
      <span
        class="mia-platform-dropdown-trigger dropdown-some-random-uuid"
      >
        <button
          class="mia-platform-btn mia-platform-btn-primary mia-platform-btn-dangerous mia-platform-btn-color-dangerous mia-platform-btn-variant-solid mia-platform-btn-icon-only mia-platform-btn-background-ghost button buttonMdIconOnly dropdownActionButton outlined"
          type="button"
        >
          <span
            class="mia-platform-btn-icon"
          >
            <svg
              aria-label="Open dropdown"
              color="currentColor"
              fill="currentColor"
              height="16"
              role="img"
              stroke="currentColor"
              stroke-width="0"
              style="color: currentColor;"
              viewBox="0 0 256 256"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M213.66,101.66l-80,80a8,8,0,0,1-11.32,0l-80-80A8,8,0,0,1,53.66,90.34L128,164.69l74.34-74.35a8,8,0,0,1,11.32,11.32Z"
              />
            </svg>
          </span>
        </button>
      </span>
    </div>
  </div>
</body>
`;

exports[`SplitButton Component ui variants renders neutral (outlined) 1`] = `
<body>
  <div>
    <div
      class="splitButton"
    >
      <button
        class="mia-platform-btn mia-platform-btn-default mia-platform-btn-color-default mia-platform-btn-variant-outlined button buttonMd mainActionButton outlined"
        type="button"
      >
        <div
          class="buttonText"
        >
          <span>
            test-trigger-button
          </span>
        </div>
      </button>
      <span
        class="mia-platform-dropdown-trigger dropdown-some-random-uuid"
      >
        <button
          class="mia-platform-btn mia-platform-btn-default mia-platform-btn-color-default mia-platform-btn-variant-outlined mia-platform-btn-icon-only button buttonMdIconOnly dropdownActionButton outlined"
          type="button"
        >
          <span
            class="mia-platform-btn-icon"
          >
            <svg
              aria-label="Open dropdown"
              color="currentColor"
              fill="currentColor"
              height="16"
              role="img"
              stroke="currentColor"
              stroke-width="0"
              style="color: currentColor;"
              viewBox="0 0 256 256"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M213.66,101.66l-80,80a8,8,0,0,1-11.32,0l-80-80A8,8,0,0,1,53.66,90.34L128,164.69l74.34-74.35a8,8,0,0,1,11.32,11.32Z"
              />
            </svg>
          </span>
        </button>
      </span>
    </div>
  </div>
</body>
`;

exports[`SplitButton Component ui variants renders primary filled 1`] = `
<body>
  <div>
    <div
      class="splitButton"
    >
      <button
        class="mia-platform-btn mia-platform-btn-primary mia-platform-btn-color-primary mia-platform-btn-variant-solid button buttonMd mainActionButton"
        type="button"
      >
        <div
          class="buttonText"
        >
          <span>
            test-trigger-button
          </span>
        </div>
      </button>
      <span
        class="mia-platform-dropdown-trigger dropdown-some-random-uuid"
      >
        <button
          class="mia-platform-btn mia-platform-btn-primary mia-platform-btn-color-primary mia-platform-btn-variant-solid mia-platform-btn-icon-only button buttonMdIconOnly dropdownActionButton"
          type="button"
        >
          <span
            class="mia-platform-btn-icon"
          >
            <svg
              aria-label="Open dropdown"
              color="currentColor"
              fill="currentColor"
              height="16"
              role="img"
              stroke="currentColor"
              stroke-width="0"
              style="color: currentColor;"
              viewBox="0 0 256 256"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M213.66,101.66l-80,80a8,8,0,0,1-11.32,0l-80-80A8,8,0,0,1,53.66,90.34L128,164.69l74.34-74.35a8,8,0,0,1,11.32,11.32Z"
              />
            </svg>
          </span>
        </button>
      </span>
    </div>
  </div>
</body>
`;

exports[`SplitButton Component ui variants renders primary outlined 1`] = `
<body>
  <div>
    <div
      class="splitButton"
    >
      <button
        class="mia-platform-btn mia-platform-btn-primary mia-platform-btn-color-primary mia-platform-btn-variant-solid mia-platform-btn-background-ghost button buttonMd mainActionButton outlined"
        type="button"
      >
        <div
          class="buttonText"
        >
          <span>
            test-trigger-button
          </span>
        </div>
      </button>
      <span
        class="mia-platform-dropdown-trigger dropdown-some-random-uuid"
      >
        <button
          class="mia-platform-btn mia-platform-btn-primary mia-platform-btn-color-primary mia-platform-btn-variant-solid mia-platform-btn-icon-only mia-platform-btn-background-ghost button buttonMdIconOnly dropdownActionButton outlined"
          type="button"
        >
          <span
            class="mia-platform-btn-icon"
          >
            <svg
              aria-label="Open dropdown"
              color="currentColor"
              fill="currentColor"
              height="16"
              role="img"
              stroke="currentColor"
              stroke-width="0"
              style="color: currentColor;"
              viewBox="0 0 256 256"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M213.66,101.66l-80,80a8,8,0,0,1-11.32,0l-80-80A8,8,0,0,1,53.66,90.34L128,164.69l74.34-74.35a8,8,0,0,1,11.32,11.32Z"
              />
            </svg>
          </span>
        </button>
      </span>
    </div>
  </div>
</body>
`;
